### PR TITLE
Changes needed for MCM OOT for Azure

### PR DESCRIPTION
**Pull request by AxiomSamarth**
**How to categorize this PR?**

/area dev-productivity
/kind api-change
/priority 4
/platform azure

**What this PR does / why we need it**:
This PR is needed for migration of AzureMachineClass to MachineClass. This migration occurs implicitly without causing rollouts of existing nodes/VMs.

**Which issue(s) this PR fixes**:
Fixes #https://github.com/gardener/machine-controller-manager-provider-azure/issues/11

**Special notes for your reviewer**:

**Release note**:

```other operator
⚠️ Before upgrading your `gardener/gardener-extension-provider-azure` to [>= v1.20.0](https://github.com/gardener/gardener-extension-provider-azure/releases/tag/v1.20.0), please upgrade your `gardener/gardener` component version to [>= v1.14.0](https://github.com/gardener/gardener/releases/tag/v1.14.0) to avoid breaking of clusters that are using the scale from/to zero feature (clusters that allowing scaling from/to 0 worker pools). If used with an older `gardener/gardener` version, this would lead to failure of clusters making use of this feature.
```

```
Migration of MCM provider from in-tree to out-of-tree. Refer - [MCM provider Azure](https://github.com/gardener/machine-controller-manager-provider-azure).
```

```
Migration of `AzureMachineClass` to `MachineClass`. This migration occurs implicitly without causing rollouts of existing nodes/VMs.
```

### CODE REVIEW

**Comment by AxiomSamarth** [comment]
reviewed/ok-to-test

File: charts/internal/machine-controller-manager/seed/values.yaml [hunk 1]
```diff
@@ -13,8 +14,9 @@ namespace:
   uid: uuid-of-namespace
 
 metricsPort: 10258
+metricsPortAWS: 10259
```
Comment (prashanth26) [inline charts/internal/machine-controller-manager/seed/values.yaml:17]:
```suggestion
metricsPortAzure: 10259
```

**Comment by AxiomSamarth** [comment]
@prashanth26 @kon-angelo @amshuman-kr

Hi All,

I have tested the extension with the new MCM-provider-Azure in the local gardener. I see things working fine. I would like to get your review comments for these changes on the Gardener Extension for Azure.

File: charts/internal/machine-controller-manager/seed/templates/deployment.yaml [hunk 2]
```diff
@@ -8,7 +8,7 @@ metadata:
     app: kubernetes
     role: machine-controller-manager
 spec:
-  revisionHistoryLimit: 1
+  revisionHistoryLimit: 0
```
Comment (prashanth26) [inline charts/internal/machine-controller-manager/seed/templates/deployment.yaml:11]:
```suggestion
  revisionHistoryLimit: 1
```
Is there a specific reason for this change? If not, we can revert this?

File: pkg/controller/worker/machines_test.go [hunk 3]
```diff
@@ -397,8 +398,22 @@ var _ = Describe("Machines", func() {
...
+								filepath.Join(azure.InternalChartsPath, "machineclass"),
+								namespace,
+								"machineclass",
+								kubernetes.Values(machineClasses),
+							),
+					)
+
 					// Test workerDelegate.DeployMachineClasses()
-					chartApplier.EXPECT().Apply(ctx, filepath.Join(azure.InternalChartsPath, "machineclass"), namespace, "machineclass", kubernetes.Values(machineClasses)).Return(nil)
+					// chartApplier.EXPECT().Apply(ctx, filepath.Join(azure.InternalChartsPath, "machineclass"), namespace, "machineclass", kubernetes.Values(machineClasses)).Return(nil)
```
Comment (prashanth26) [inline pkg/controller/worker/machines_test.go:416]:
```suggestion
					chartApplier.EXPECT().Apply(ctx, filepath.Join(azure.InternalChartsPath, "machineclass"), namespace, "machineclass", kubernetes.Values(machineClasses)).Return(nil)
```
Why is this commented?

**Comment by prashanth26** [comment]
Also, you can probably update the release notes to the following.

```breaking operator
⚠️ Before upgrading your `gardener/gardener-extension-provider-azure` to [>= v1.19.0](https://github.com/gardener/gardener-extension-provider-azure/releases/tag/v1.19.0), please upgrade your `gardener/gardener` component version to [>= v1.14.0](https://github.com/gardener/gardener/releases/tag/v1.14.0) to avoid breaking of clusters that are using the scale from/to zero feature (clusters that allowing scaling from/to 0 worker pools). If used with an older `gardener/gardener` version, this would lead to failure of clusters making use of this feature. 
```
```other developer
Migration of MCM provider from in-tree to out-of-tree. Refer - [MCM provider Azure](https://github.com/gardener/machine-controller-manager-provider-azure).
```
```other developer
Migration of `AzureMachineClass` to `MachineClass`. This migration occurs implicitly without causing rollouts of existing nodes/VMs.
```

File: charts/internal/machineclass/templates/machineclass.yaml [hunk 4]
```diff
@@ -5,81 +5,39 @@ kind: Secret
 metadata:
   name: {{ $machineClass.name }}
   namespace: {{ $.Release.Namespace }}
-{{- if $machineClass.labels }}
+{{- if $machineClass.secret.labels }}
```
Comment (kon-angelo) [inline charts/internal/machineclass/templates/machineclass.yaml]: The `secret` here doesn't seem correct. These are labels for the secret.

File: charts/internal/machineclass/templates/machineclass.yaml [hunk 5]
```diff
@@ -5,81 +5,39 @@ kind: Secret
...
   labels:
-{{ toYaml $machineClass.labels | indent 4 }}
+{{ toYaml $machineClass.secret.labels | indent 4 }}
```
Comment (kon-angelo) [inline charts/internal/machineclass/templates/machineclass.yaml]: Same as previous comment.

File: charts/images.yaml [hunk 6]
```diff
@@ -32,7 +32,10 @@ images:
   sourceRepository: github.com/gardener/machine-controller-manager
   repository: eu.gcr.io/gardener-project/gardener/machine-controller-manager
   tag: "v0.37.0"
-
+- name: machine-controller-manager-provider-azure
+  sourceRepository: github.com/gardener/machine-controller-manager-provider-azure
+  repository: eu.gcr.io/gardener-project/gardener/machine-controller-manager-provider-azure
+  tag: "0.1.0-dev-6ed01a9cab7390c3147b2da7c1c132b84b9e8fb5"
```
[the remark this answers is not in the document]
Comment (AxiomSamarth) [reply charts/images.yaml:38]: I will keep it mind, Konstantinos. Thank you :)

**Comment by prashanth26** [comment]
Overall changes looks good to me. Can be merged post releasing of https://github.com/gardener/machine-controller-manager-provider-azure

**Comment by AxiomSamarth** [comment]
> Overall changes looks good to me. Can be merged post releasing of https://github.com/gardener/machine-controller-manager-provider-azure

@gardener/gardener-extension-provider-azure-maintainers 

[MCM-provider-Azure](https://github.com/gardener/machine-controller-manager-provider-azure/releases/tag/0.1.0) has been released. I have integrated the changes here and tested in my local Gardener setup too. Everything works as expected. Would you please take a look at the changes too and provide the feedback/review comments if any. So, that we can have the MCM-Provider-OOT deployed on the landscapes in the near future 😄

**Comment by AxiomSamarth** [comment]
I think that the version file entry should have been "v0.1.0" instead of "0.1.0" in MCM-provider-azure release. https://github.com/gardener/machine-controller-manager-provider-azure/blob/0.1.0/VERSION here . How do we go about for this?

**Comment by ialidzhikov** [comment]
> I think that the version file entry should have been "v0.1.0" instead of "0.1.0" in MCM-provider-azure release. https://github.com/gardener/machine-controller-manager-provider-azure/blob/0.1.0/VERSION here . How do we go about for this?

You can use for now `0.1.0`. For future releases - https://github.com/gardener/machine-controller-manager-provider-azure/pull/18.

**Comment by AxiomSamarth** [comment]
> > I think that the version file entry should have been "v0.1.0" instead of "0.1.0" in MCM-provider-azure release. https://github.com/gardener/machine-controller-manager-provider-azure/blob/0.1.0/VERSION here . How do we go about for this?
> 
> You can use for now `0.1.0`. For future releases - [gardener/machine-controller-manager-provider-azure#18](https://github.com/gardener/machine-controller-manager-provider-azure/pull/18).

Thank you, @ialidzhikov and yes, I have made a note of the prefix for the future releases and as well merged your PR with the prefix already :)

File: pkg/controller/worker/machines.go [hunk 7]
```diff
@@ -62,6 +64,12 @@ func (w *workerDelegate) DeployMachineClasses(ctx context.Context) error {
 			return err
 		}
 	}
+
+	// Delete any older version of AzureMachineClass CRs.
```
Comment (vpnachev) [inline pkg/controller/worker/machines.go:68]:
```suggestion
	// Delete any older version of AzureMachineClass CRs.
	// TODO: Remove this clean-up in future version.
```

File: pkg/controller/worker/machines_test.go [hunk 8]
```diff
@@ -25,6 +25,7 @@ import (
 	apiv1alpha1 "github.com/gardener/gardener-extension-provider-azure/pkg/apis/azure/v1alpha1"
 	"github.com/gardener/gardener-extension-provider-azure/pkg/azure"
 	. "github.com/gardener/gardener-extension-provider-azure/pkg/controller/worker"
+	"sigs.k8s.io/controller-runtime/pkg/client"
```
Comment (vpnachev) [inline pkg/controller/worker/machines_test.go:28]: Move this import to the 3rd-party modules import block.

File: pkg/controller/worker/machines_test.go (hunk 3 above)
Comment (vpnachev) [inline pkg/controller/worker/machines_test.go:416]:
```suggestion
```

File: charts/internal/machine-controller-manager/seed/templates/deployment.yaml [hunk 9]
```diff
@@ -39,24 +39,51 @@ spec:
       serviceAccountName: machine-controller-manager
       terminationGracePeriodSeconds: 5
       containers:
-      - name: azure-machine-controller-manager
-        image: {{ index .Values.images "machine-controller-manager" }}
+      - name: machine-controller-manager-provider-azure
```
Comment (vpnachev) [inline charts/internal/machine-controller-manager/seed/templates/deployment.yaml:42]: You need to extend the logging configuration here https://github.com/gardener/gardener-extension-provider-azure/blob/master/charts/gardener-extension-provider-azure/templates/configmap-logging.yaml to collect properly the logs from the newly added container

Comment (prashanth26) [reply charts/internal/machine-controller-manager/seed/templates/deployment.yaml:42]: Hi @vpnachev ,

With the previous PRs for migration to OOT for other providers, I didn't have to make changes to collect logs from the new containers. My guess is that [this configuration](https://github.com/gardener/gardener-extension-provider-azure/blob/master/charts/gardener-extension-provider-azure/templates/configmap-logging.yaml#L19) already collects the logs?

Refer AWS log collection templates as well [here](https://github.com/gardener/gardener-extension-provider-aws/blob/master/charts/gardener-extension-provider-aws/templates/configmap-logging.yaml).

Comment (vpnachev) [reply charts/internal/machine-controller-manager/seed/templates/deployment.yaml:42]: This configuration https://github.com/gardener/gardener-extension-provider-azure/blob/4159e890852a5ccc9020502aba7953be6567dd4d/charts/gardener-extension-provider-azure/templates/configmap-logging.yaml#L19 matches only the other container 

/invite @Kristian-ZH @vlvasilev

File: VERSION [hunk 10]
```diff
@@ -1 +1 @@
-v1.19.0-dev
+v1.19.0-dev
```
Comment (ialidzhikov) [inline VERSION:1]: You could revert the change in the VERSION file.

File: charts/internal/machine-controller-manager/seed/templates/deployment.yaml [hunk 11]
```diff
@@ -86,4 +113,4 @@ spec:
       volumes:
       - name: machine-controller-manager
         secret:
-          secretName: machine-controller-manager
+          secretName: machine-controller-manager
```
Comment (ialidzhikov) [inline charts/internal/machine-controller-manager/seed/templates/deployment.yaml:116]: You could add an empty line to the end of the file.

File: charts/internal/machine-controller-manager/seed/values.yaml [hunk 12]
```diff
@@ -13,8 +14,9 @@ namespace:
   uid: uuid-of-namespace
 
 metricsPort: 10258
+metricsPortAzure: 10259
 
 vpa:
   enabled: true
   updatePolicy:
-    updateMode: "Auto"
+    updateMode: "Auto"
```
Comment (ialidzhikov) [inline charts/internal/machine-controller-manager/seed/values.yaml:22]: You could add an empty line to the end of the file.

File: charts/internal/machineclass/templates/machineclass.yaml [hunk 13]
```diff
@@ -82,4 +76,11 @@ spec:
...
 {{- end }}
-{{- end }}
+secretRef:
+  name: {{ $machineClass.name }}
+  namespace: {{ $.Release.Namespace }}
+credentialsSecretRef:
+  name: {{ $machineClass.credentialsSecretRef.name }}
+  namespace: {{ $machineClass.credentialsSecretRef.namespace }}
+provider: "Azure"
+{{- end }}
```
Comment (ialidzhikov) [inline charts/internal/machineclass/templates/machineclass.yaml:86]: You could add an empty line to the end of the file.

**Review by ialidzhikov — COMMENTED** [review]
Nits

**Comment by AxiomSamarth** [comment]
I need some help to fix this build error. When I run `make revendor` I get the below error

```
warning: ignoring symlink /Users/i538135/go/src/github.com/gardener/gardener-extension-provider-azure/charts/internal/seed-controlplane/charts/cloud-controller-manager/charts/utils-tls-cipher-suites
github.com/gardener/gardener-extension-provider-azure/hack imports
        github.com/gardener/gardener/.github: malformed import path "github.com/gardener/gardener/.github": leading dot in path element
github.com/gardener/gardener-extension-provider-azure/hack imports
        github.com/gardener/gardener/.github/ISSUE_TEMPLATE: malformed import path "github.com/gardener/gardener/.github/ISSUE_TEMPLATE": leading dot in path element
github.com/gardener/gardener-extension-provider-azure/hack imports
        github.com/gardener/gardener/hack/.ci: malformed import path "github.com/gardener/gardener/hack/.ci": leading dot in path element
make: *** [revendor] Error 1
```

**Comment by ialidzhikov** [comment]
> I need some help to fix this build error. When I run `make revendor` I get the below error
> 
> ```
> warning: ignoring symlink /Users/username/go/src/github.com/gardener/gardener-extension-provider-azure/charts/internal/seed-controlplane/charts/cloud-controller-manager/charts/utils-tls-cipher-suites
> github.com/gardener/gardener-extension-provider-azure/hack imports
>         github.com/gardener/gardener/.github: malformed import path "github.com/gardener/gardener/.github": leading dot in path element
> github.com/gardener/gardener-extension-provider-azure/hack imports
>         github.com/gardener/gardener/.github/ISSUE_TEMPLATE: malformed import path "github.com/gardener/gardener/.github/ISSUE_TEMPLATE": leading dot in path element
> github.com/gardener/gardener-extension-provider-azure/hack imports
>         github.com/gardener/gardener/hack/.ci: malformed import path "github.com/gardener/gardener/hack/.ci": leading dot in path element
> make: *** [revendor] Error 1
> ```

Update your local go version to the latest one.

**Comment by AxiomSamarth** [comment]
Urgggh! Something just got messed. I will clean it up and open a fresh PR!

**Comment by vpnachev** [comment]
> Urgggh! Something just got messed. I will clean it up and open a fresh PR!

Please no! It is very hard for anybody to keep reviewing the same change over multiple PRs.
Please fix your local branch/fork and push to this PR.

**Comment by AxiomSamarth** [comment]
> > Urgggh! Something just got messed. I will clean it up and open a fresh PR!
> 
> Please no! It is very hard for anybody to keep reviewing the same change over multiple PRs.
> Please fix your local branch/fork and push to this PR.

Okayy!

File: charts/gardener-extension-provider-azure/templates/configmap-logging.yaml [hunk 14]
```diff
@@ -16,7 +16,7 @@ data:
 
     [FILTER]
         Name                parser
-        Match               kubernetes.*machine-controller-manager*azure-machine-controller-manager*
+        Match               kubernetes.*machine-controller-manager*machine-controller-manager-provider-azure*
```
Comment (prashanth26) [inline charts/gardener-extension-provider-azure/templates/configmap-logging.yaml:19]:
```suggestion
        Match               kubernetes.*azure-machine-controller-manager*machine-controller-manager-provider-azure*
```
Shouldn't this match with [this](https://github.com/gardener/gardener-extension-provider-azure/pull/257/files#diff-e38adde4c814c39f0c3443b395a76dfc43639b003e8514c731b39e2b1683aea9R71)?

Comment (AxiomSamarth) [reply charts/gardener-extension-provider-azure/templates/configmap-logging.yaml:19]: I am not sure. But [this comment](https://github.com/gardener/gardener-extension-provider-azure/pull/257#discussion_r607122158) from Vladimir suggested to extend it to ConfigMap-logging. Did I comprehend it wrongly by any chance?

Comment (prashanth26) [reply charts/gardener-extension-provider-azure/templates/configmap-logging.yaml:19]: Your understanding is right. However both the container names must be matching right. Else we may end up not collecting logs from the other/old container.

Comment (AxiomSamarth) [reply charts/gardener-extension-provider-azure/templates/configmap-logging.yaml:19]: Got it. Got it! Thanks @prashanth26

**Comment by AxiomSamarth** [comment]
/invite @vpnachev @vlvasilev @Kristian-ZH

Hi All, I have made all the suggested changes. I request a review from you all, please.

cc: @prashanth26

File: charts/gardener-extension-provider-azure/templates/configmap-logging.yaml [hunk 15]
```diff
@@ -16,7 +16,7 @@ data:
 
     [FILTER]
         Name                parser
-        Match               kubernetes.*machine-controller-manager*azure-machine-controller-manager*
+        Match               kubernetes.*azure-machine-controller-manager*machine-controller-manager-provider-azure*
```
Comment (Kristian-ZH) [inline charts/gardener-extension-provider-azure/templates/configmap-logging.yaml:19]: The regex here has the following syntax:
kubernetes.*<pod-name-prefix>*<container-name>*

As I see the pod name is still `machine-controller-manager`, so this change will no work.

Comment (Kristian-ZH) [reply charts/gardener-extension-provider-azure/templates/configmap-logging.yaml:19]: Also now, when there is a second container in the deployment, you should create a new filter which caches it.
You can check the log format of the new container. If it is the same as the other one, you can use the same Parser.
Otherwise I will help you to create a new one

Comment (AxiomSamarth) [reply charts/gardener-extension-provider-azure/templates/configmap-logging.yaml:19]: So, @Kristian-ZH what is the right thing to put here?

Comment (Kristian-ZH) [reply charts/gardener-extension-provider-azure/templates/configmap-logging.yaml:19]: Can you please check the log format of the new container via `kubectl logs machine-controller-manager -c azure-machine-controller-manager` and to compare it with the one from `kubectl logs machine-controller-manager machine-controller-manager-provider-azure`

Comment (prashanth26) [reply charts/gardener-extension-provider-azure/templates/configmap-logging.yaml:19]: The old container here is `azure-machine-controller-manager` and the new container is `machine-controller-manager machine-controller-manager-provider-azure`. Also I don't think there should be any change in logging format as we have just split the single MCM container into two containers now. @AxiomSamarth you can confirm the same by looking at the logs. 

The concern here is the Match filters that are to be used to get the container logs from both containers.

Comment (AxiomSamarth) [reply charts/gardener-extension-provider-azure/templates/configmap-logging.yaml:19]: I just checked. I do not see any difference in the logging format.

Comment (Kristian-ZH) [reply charts/gardener-extension-provider-azure/templates/configmap-logging.yaml:19]: okay, then for now, the Filter an be changed to:
```
    [FILTER]
        Name                      parser
        Match                      kubernetes.*machine-controller-manager*azure*
        Key_Name              log
        Parser                     kubeapiserverParser
        Reserve_Data        True
```
This will catch all containers from the machine-controller-manager pod which contains `azure` in their names.
In the future if new containers are added to the pod, we can split this filter to dedicated filters per container.

Comment (AxiomSamarth) [reply charts/gardener-extension-provider-azure/templates/configmap-logging.yaml:19]: Thank you @Kristian-ZH

Comment (prashanth26) [reply charts/gardener-extension-provider-azure/templates/configmap-logging.yaml:19]: Okay thanks for the comments and feedback.

**Review by Kristian-ZH — DISMISSED** [review]

**Comment by prashanth26** [comment]
@vpnachev - I think @AxiomSamarth has made all the changes now. If it looks good, can we merge this PR?

**Review by kon-angelo — APPROVED** [review]
/lgtm
